org.eclipse.search/search/org/eclipse/search/internal/ui/text/FileTreeContentProvider.java

```diff
@@ -177,16 +177,19 @@ private boolean hasMatches(Object element) {
 		if (element instanceof LineElement) {
 			LineElement lineElement= (LineElement) element;
 			IResource resource = lineElement.getParent();
-			if (getMatchCount(resource) > 0) {
+			if (resourceHasMatches(resource)) {
 				return lineElement.hasMatches(fResult);
 			}
 		}
 		return fPage.getDisplayedMatchCount(element) > 0;
 	}
 
-	private int getMatchCount(Object element) {
-		return fResult.getActiveMatchFilters().length > 0 ? fPage.getDisplayedMatchCount(element)
-				: fResult.getMatchCount();
+	private boolean resourceHasMatches(IResource element) {
+		if (fResult.getActiveMatchFilters().length > 0) {
+			return fPage.getDisplayedMatchCount(element) > 0;
```

Comment:
Note: I haven't debugged and checked if this branch need similar fix.
`fPage.getDisplayedMatchCount(element) > 0` is called twice and could be replaced with `fPage.hasMatches(element)` that could be implemented in a way that doesn't require calculating everything.